### PR TITLE
[Workflow] Public function to get possible transitions in base of the configuration

### PR DESCRIPTION
**Pull request by ruudy-es**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Allow new functionality to know the possible transitions based on the current marking. Just a dynamic recognition of the possible transitions configured, no "can" function being checked to avoid launch guard events multiple times.

### CODE REVIEW

File: src/Symfony/Component/Workflow/Tests/WorkflowTest.php [hunk 1]
```diff
@@ -226,6 +226,27 @@ public function testGetEnabledTransitions()
         $this->assertSame('t5', $transitions['t5']->getName());
     }
 
+    public function teestGetPossibleTransitions()
```
Comment (patrick-mcdougle) [inline src/Symfony/Component/Workflow/Tests/WorkflowTest.php]: typo

**Comment by lyrixx** [comment]
There is already the `getEnabledTransitions` method. Indeed it triggers the guard event. It's required because the guard event can deny a transition.
So for now I don't understand why you need this method and so I'm :-1:

**Comment by ruudy-es** [comment]
That´s exactly the problem, based on the existing Workflow class the Guards are being launched on following functions:

- getEnabledTransitions
- can
- apply

So, if you use your Guards to validate somehow required rules to apply the transition, that validation will be launched many times, on the example from Paris you have to include the apply inside a try{}catch(){} to avoid multiple checks on Guards.

Imagine the scenario where i just want to get the configured transitions, iterate over them and throw the can / apply to automatise the Workflow, this is possible with the getEnabledTransitions method, but the Guards will be launched twice.

For me, both methods (getEnabled / get Possibles) cover different scenarios, anyway, based on this Workflow class where every method call the doCan method, no one will be able to avoid the Guards, you do not need to force the people to throw the Guards on getEnabledTransitions, it will be launched on apply...

**Comment by lyrixx** [comment]
I don't see what is the problem of calling many time guard event listeners. Each time they are called with different arguments. So the decision can be different according to the argument. 
If you don't care of "where" the marking is and you take a decision only on the transition, you can add a cache layer if you think it's a performance culprit.

Anyway, as you said, listeners will be called anyway during the apply / can.

Finally, I really like the simplicity of the Workflow API (getEnabledTransitions / can / apply). If we add this method, it could confuse new comers. 

For theses reason, I'm still :-1: for this PR

But If you are able to give us a reproducer of your issue, It will be easier to figure out what's the real issue.

**Comment by javiereguiluz** [comment]
@ruudy-es thanks for proposing this improvement but sadly I need to close this PR as "won't fix" because of the `-1` votes from the Symfony Core Team members.

The Workflow component is very recent, so we are very cautious about adding new features to it. In any case, if more users complain about this very same thing, we'll reconsider your proposal for Symfony 3.3. Thanks!

**Pull request closed without merging**